File: go/sample_02_otp.go
/*
Sample 02 — Yêu cầu & Xác thực OTP (Request OTP & Verify OTP / Smart OTP)
========================================================================
Hướng dẫn luồng Request OTP và Polling Smart OTP Push Notification.
*/

package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/SSI-Securities-Inc/ssi-sdk-go/v3/ssi"
)

func main() {
	config, err := GetConfig()
	if err != nil {
		log.Fatalf("Không đọc được cấu hình: %v", err)
	}

	auth := ssi.NewAuth(config)
	defer auth.Close()

	fmt.Println("=== Bước 1: Yêu cầu OTP (Request OTP) ===")
	otpRes, err := auth.RequestOTP()
	if err != nil {
		log.Fatalf("Lỗi Request OTP: %v", err)
	}
	fmt.Printf("Request OTP Response: %+v\n", otpRes)

	var transactionID string
	if dataMap, ok := otpRes["data"].(map[string]interface{}); ok {
		if tid, ok := dataMap["transactionId"].(string); ok {
			transactionID = tid
		}
	}
	if transactionID == "" {
		if tid, ok := otpRes["transactionId"].(string); ok {
			transactionID = tid
		}
	}

	if transactionID != "" {
		fmt.Printf("\n[Smart OTP] Đã nhận Transaction ID: %s\n", transactionID)
		fmt.Println("Vui lòng mở ứng dụng SSI trên điện thoại và bấm APPROVE (Xác nhận)...")
		fmt.Println("SDK đang Polling chờ bạn bấm phê duyệt...")

		accessToken, err := auth.EnsureAuthenticated("", transactionID)
		if err != nil {
			log.Fatalf("\n[LỖI/TIMEOUT] Phê duyệt Smart OTP thất bại: %v", err)
		}

		fmt.Println("\n[THÀNH CÔNG] Đã xác thực Smart OTP!")
		fmt.Printf("Access Token: %s...\n", accessToken[:minInt(40, len(accessToken))])
	} else {
		fmt.Print("\n[OTP Thường / Smart OTP lấy trực tiếp trên App] Nhập mã OTP 6 số: ")
		reader := bufio.NewReader(os.Stdin)
		userOTP, _ := reader.ReadString('\n')
		userOTP = strings.TrimSpace(userOTP)

		if userOTP != "" {
			token, err := auth.Authenticate(userOTP)
			if err != nil {
				log.Fatalf("Lỗi xác thực OTP: %v", err)
			}
			fmt.Println("\n[THÀNH CÔNG] Đã xác thực mã OTP!")
			fmt.Printf("Access Token: %s...\n", token.AccessToken[:minInt(40, len(token.AccessToken))])
		}
	}
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: dotnet/Sample02Otp.cs
using System.Text.Json;
using SsiSdk;

/// <summary>
/// Sample 02 — Yêu cầu & Xác thực OTP (Request OTP & Verify OTP / Smart OTP)
/// </summary>
static class Sample02Otp
{
    public static async Task RunAsync()
    {
        var config = SampleConfig.Create();
        using var auth = new AuthClient(config);

        Console.WriteLine("=== Bước 1: Yêu cầu OTP (Request OTP) ===");
        var otpRes = await auth.RequestOtpAsync();
        Console.WriteLine($"Request OTP Response: {otpRes}");

        string? transactionId = null;
        if (otpRes.ValueKind == JsonValueKind.Object)
        {
            if (otpRes.TryGetProperty("data", out var dataObj) && dataObj.ValueKind == JsonValueKind.Object)
            {
                if (dataObj.TryGetProperty("transactionId", out var tid))
                    transactionId = tid.GetString();
            }
            if (string.IsNullOrEmpty(transactionId) && otpRes.TryGetProperty("transactionId", out var tid2))
            {
                transactionId = tid2.GetString();
            }
        }

        if (!string.IsNullOrEmpty(transactionId))
        {
            Console.WriteLine($"\n[Smart OTP] Đã nhận Transaction ID: {transactionId}");
            Console.WriteLine("Vui lòng mở ứng dụng SSI trên điện thoại và bấm APPROVE (Xác nhận)...");
            Console.WriteLine("SDK đang Polling chờ bạn bấm phê duyệt...");

            try
            {
                var accessToken = await auth.EnsureAuthenticatedAsync(
                    transactionId: transactionId,
                    pollInterval: TimeSpan.FromSeconds(5),
                    pollMaxRetries: 6
                );
                Console.WriteLine("\n[THÀNH CÔNG] Đã xác thực Smart OTP!");
                Console.WriteLine($"Access Token: {accessToken[..Math.Min(40, accessToken.Length)]}...");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n[LỖI/TIMEOUT] Phê duyệt Smart OTP thất bại: {ex.Message}");
            }
        }
        else
        {
            Console.Write("\n[OTP Thường / Smart OTP lấy trực tiếp trên App] Vui lòng nhập mã OTP 6 số: ");
            var userOtp = Console.ReadLine()?.Trim() ?? "";
            if (!string.IsNullOrEmpty(userOtp))
            {
                var token = await auth.AuthenticateAsync(otp: userOtp);
                Console.WriteLine("\n[THÀNH CÔNG] Đã xác thực mã OTP!");
                Console.WriteLine($"Access Token: {token.AccessToken[..Math.Min(40, token.AccessToken.Length)]}...");
            }
        }
    }
}
